### ISSUE TITLE
[RFE] Report measurements also when job times out

### ISSUE THREAD

**Issue by ygalblum** [kube-burner/kube-burner#742]
**Is your feature request related to a problem? Please describe.**

I would like to use `kube-burner` for the following test:
The test contains a single `create` job for creating VMs with multiple iterations and in each iteration create multiple replicas until the cluster is unable to start any more VMs. Then, I want to gather the measurements to understand the performance implications of each iteration. We would like to examine the measurements of each iteration, so we need to wait for all VMs to become running before moving to the next iteration.

Currently this cannot be achieved for two reasons:
1. There is no infinite option to `jobIteration`. This can be overcome by setting a value high enough so that the last iteration will never be reached
2. The test requires waiting for all objects in each iteration. However, currently, if a timeout occurs when waiting for the objects, `kube-burner` returns immediately without collecting the measurements. This issue comes to discuss this reason.

**Describe the solution you'd like**
Currently, the wait method calls `log.Fatalf` which prints the message and exists non-zero. I would like to change the behavior so that the error is propagated to the caller, letting it decide what to do next.

Then, we need to decide what comes next. Should the processing continue? If yes, then at what level - move to the next iteration? move to the next job? If not, should it just report the measurements are exit non-zero?

I think the latter is the more expected behavior. That is, the run should stop and fail. But, it should still produce the measurements. Then the users can decide whether the measurements are useful for them or not

**Describe alternatives you've considered**
I see two options:
1. Manual iteration - run only one iteration until the test fails. This way all measurements are collected, but separately.
2. Duplicate the job instead of iterations. But, how many times?
I think the second option is an OK fallback. But, I'd rather see a complete test that can be executed in a single call.

**Comment by ygalblum** [comment]
@rsevilla87 thought?

**Comment by rsevilla87** [comment]
Hey @ygalblum , some thoughts.

1. With respect of the number of iterations, you're right, there's not such parameter of "infinite", in order to achieve such behaviour in some of our workloads we're using some additional logic in the some of the workloads available in [kube-burner-ocp](https://github.com/kube-burner/kube-burner-ocp/blob/main/node-density.go#L45), in fact, I've a WIP workload that deploys VMs in the given cluster that uses a similar approach to "fill" the worker nodes with VMs, find it at https://github.com/kube-burner/kube-burner-ocp/pull/154

 2. There's a parameter `podWait` (naming can be confusing), that makes kube-burner to wait for objects in a per-iteration basis, rather than a job basis.
 
 4. As you mention, when the wait period times out (defined by MaxWaitTimeout) , kube-burner exits the benchmark execution with RC 1 and a Fatal log message, I think it can be useful to extend this behaviour and provide the user the option of what to do when a waiter times out, something like `waitTimeoutAction`.
 
 6. Measurements collection are not tied to waiters, but as the current behavior is to hang, measurement data is not collected and  reported/indexed in any case.

Let me know what you think

**Comment by ygalblum** [comment]
1. My target test is more elaborate. Meaning, it has additional steps for each iteration of VM creation such as storage resize and migration. But, even if I just look at this test, it will get to the same issue of metrics not being reported upon failure.
2. I'm using `podWait`. This is where the run fails and returns without providing the data
3. Sure, there's no harm in doing that. But, I think it is separate from providing the metrics
4. Yes, this is the main issue. The fact that upon timeout the `kube-burner` exits immediately is the root cause of my issue.

To summarize, what I would like to do is replace the `Fatalf` exits to error returns and let the caller decide what to do. This will allow the executor to report the metrics before exiting.

**Comment by ygalblum** [comment]
Not sure id adding a comment with remove the tag. But, worth a try

**Comment by vaidikcode** [comment]
Maybe We can implement a feature where using `waitTimeoutAction` user specifies either to `collect-then-exit`(which means collect matrics just before a job fails and exits) and `fast-fail`(which is the current default behavior). Also rather than returning `log.fatalf` and exiting, we return errors so that using the user-specified fields we can handle the required behavior. wdyt ? @ygalblum @rsevilla87

**Comment by vishnuchalla** [comment]
Initial PR capturing possible fatals before timeout: https://github.com/kube-burner/kube-burner/pull/1044. We need follow up that acts based on  `waitTimeoutAction` flag. I am thinking of just having it as a boolean flag to capture all config declared measurements if enabled as true. 
cc: @rsevilla87 @ygalblum

**Comment by afcollins** [comment]
I'm not sure this should be closed as it is not done.

I just ran into the same pain point on kube-burner-ocp v1.11.2. The workload breached the kube-burner timeout, and I lost all the podLatencyMeasurements from 4 hours of running.

We need to write at least kube-burner measurements locally (if indexing locally) since those are stored in-memory to kube-burner. Any metrics we can at least scrape from prometheus after the fact. Any Measurements are lost forever and we have to run again.

**Comment by vishnuchalla** [comment]
@afcollins lets take this in a follow up https://github.com/kube-burner/kube-burner/issues/1155. Resolution for this issue enabled us to capture all the errors during multiple phases of execution. But that being a case at multiple places in our code base and not clear on which metrics to capture exactly at that point in time, we have closed it deciding to take more speicific extensions in a follow up

**Comment by afcollins** [comment]
OK, sure.
